control-plane/api/v1alpha1/api_gateway_types.go

```diff
@@ -0,0 +1,260 @@
+	// More Info: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// Annotation Information to copy to services or deployments
+	CopyAnnotations CopyAnnotationsSpec `json:"copyAnnotations,omitempty"`
+}
+
+// PortSpec defines the ports that the API Gateway will listen on.
+type PortSpec struct {
+	// The port for Consul's HTTP server.
+	HTTP int `json:"http,omitempty"`
+	// The grpc port for Consul's xDS server.
+	GRPC int `json:"grpc,omitempty"`
+}
+
+//+kubebuilder:object:generate=true
+
+// CopyAnnotationsSpec defines the annotations that should be copied to the gateway service.
+type CopyAnnotationsSpec struct {
+	// List of annotations to copy to the gateway service.
+	Service []string `json:"service,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// GatewayClassConfigList is a list of Config resources.
+type GatewayClassConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is the list of Configs.
+	Items []GatewayClassConfig `json:"items"`
+}
+
+// RoleFor constructs a Kubernetes Role for the specified Gateway based
+// on the GatewayClassConfig.
+func (c *GatewayClassConfig) RoleFor(gw *gwv1beta1.Gateway) *rbac.Role {
+	return &rbac.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      gw.Name,
+			Namespace: gw.Namespace,
+			Labels:    labelsForGateway(gw),
+		},
+		Rules: []rbac.PolicyRule{{
+			APIGroups: []string{"policy"},
+			Resources: []string{"podsecuritypolicies"},
+			Verbs:     []string{"use"},
+		}},
+	}
+}
+
+// RoleBindingFor constructs a Kubernetes RoleBinding for the specified Gateway
+// based on the GatewayClassConfig. If the GatewayClassConfig is configured in
+// such a way that does not require a RoleBinding, nil is returned.
+func (c *GatewayClassConfig) RoleBindingFor(gw *gwv1beta1.Gateway) *rbac.RoleBinding {
+	serviceAccount := c.ServiceAccountFor(gw)
+	if serviceAccount == nil {
+		return nil
+	}
+
+	role := c.RoleFor(gw)
+	if role == nil {
+		return nil
+	}
+
+	return &rbac.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      gw.Name,
+			Namespace: gw.Namespace,
+			Labels:    labelsForGateway(gw),
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "Role",
+			Name:     role.Name,
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      serviceAccount.Name,
+				Namespace: serviceAccount.Namespace,
+			},
+		},
+	}
+}
+
+// ServiceAccountFor constructs a Kubernetes ServiceAccount for the specified
+// Gateway based on the GatewayClassConfig.
+func (c *GatewayClassConfig) ServiceAccountFor(gw *gwv1beta1.Gateway) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      gw.Name,
+			Namespace: gw.Namespace,
+			Labels:    labelsForGateway(gw),
+		},
+	}
+}
+
+func MergeSecret(a, b *corev1.Secret) *corev1.Secret {
```

Comment:
actually pretty much for all of the helpers below -- I think we should add them as we find the need for them throughout the subsequent dev process. Think you can drop pretty much all of the helpers after the initial struct definitions.